R/assign_portfolio.R

```diff
@@ -21,63 +21,144 @@
+
+  # Exit condition for identical sorting variables
+  if (length(unique(sorting_values)) == 1) {
+    cli::cli_warn("The sorting variable is constant and only one portfolio is returned.")
+    return(rep(1, nrow(data_breakpoints)))
+  }
+
   breakpoints <- quantile(
     sorting_values, probs = probs, na.rm = TRUE, names = FALSE
   )
 
+  if (fix_extremes) {
+    # Portfolio 1 and n are overpopulated
+    if (breakpoints[1] == breakpoints[2] && breakpoints[n_portfolios] == breakpoints[n_portfolios + 1]) {
+      if (!is.null(percentiles)) {
+        cli::cli_warn("`fix_extremes` is TRUE and equally-spaced portfolios are returned for non-edge portfolios.")
+      }
+
+      sorting_values_new <- sorting_values[which(sorting_values > breakpoints[1] & sorting_values < breakpoints[n_portfolios + 1])]
+
+      probs_new <- seq(0, 1, length.out = n_portfolios - 1)
+
+      breakpoints_new <- quantile(
+        sorting_values_new, probs = probs_new, na.rm = TRUE, names = FALSE
+      )
+
+      breakpoints_new[n_portfolios - 1] <- breakpoints_new[n_portfolios - 1] + 1e-15
+
+      breakpoints <- c(breakpoints[1], breakpoints_new, breakpoints[n_portfolios + 1])
+    }
+
+    # Portfolio 1 is overpopulated
+    if (breakpoints[1] == breakpoints[2]) {
+      if (!is.null(percentiles)) {
+        cli::cli_warn("`fix_extremes` is TRUE and equally-spaced portfolios are returned for non-edge portfolios.")
+      }
+
+      sorting_values_new <- sorting_values[which(sorting_values > breakpoints[1])]
+
+      probs_new <- seq(0, 1, length.out = n_portfolios)
+
+      breakpoints_new <- quantile(
+        sorting_values_new, probs = probs_new, na.rm = TRUE, names = FALSE
+      )
+
+      breakpoints <- c(breakpoints[1], breakpoints_new)
+    }
+
+    # Portfolio n is overpopulated
+    if (breakpoints[n_portfolios] == breakpoints[n_portfolios + 1]) {
+      if (!is.null(percentiles)) {
+        cli::cli_warn("`fix_extremes` is TRUE and equally-spaced portfolios are returned for non-edge portfolios.")
+      }
+
+      sorting_values_new <- sorting_values[which(sorting_values < breakpoints[n_portfolios])]
+
+      probs_new <- seq(0, 1, length.out = n_portfolios)
+
+      breakpoints_new <- quantile(
+        sorting_values_new, probs = probs_new, na.rm = TRUE, names = FALSE
+      )
+
+      breakpoints_new[n_portfolios] <- breakpoints_new[n_portfolios] + 1e-15
+
+      breakpoints <- c(breakpoints_new, breakpoints[n_portfolios + 1])
+    }
+  }
+
+  breakpoints[2:(n_portfolios + 1)] <- breakpoints[2:(n_portfolios + 1)] + 1e-20
+
+  # Assign portfolios
   sorting_values_all <- data[[sorting_variable]]
   portfolio_indices <- findInterval(
     sorting_values_all, breakpoints, all.inside = TRUE
   )
 
-  portfolio_indices
+  if (length(unique(portfolio_indices)) != n_portfolios) {
+    cli::cli_warn("The number of portfolios differs from the specified parameter due to large clusters in the sorting variable.")
```

Comment:
If the sorting variable contains NA values portfolio_indices will return NAs for these observations too and thus the condition may be true even without clusters in the sorting variable. 
length(unique(na.omit(portfolio_indices)) != n_portfolios may do the job (I guess that an additional warning regarding NA in the sorting variable may be useful in the beginning.